packages/tgsl-tools/src/parsers.ts

```diff
@@ -259,62 +259,74 @@ export type TranspilationResult = {
   externalNames: string[];
 };
 
-export function extractFunctionParts(program: acorn.Program): {
+export function extractFunctionParts(rootNode: acorn.AnyNode): {
   params: acorn.Identifier[];
   body: acorn.BlockStatement | acorn.Expression;
 } {
-  const programBody = program.body[0];
-
-  if (
-    programBody?.type === 'ExpressionStatement' &&
-    programBody.expression.type === 'ArrowFunctionExpression'
-  ) {
-    const mainExpression = programBody.expression;
-
-    if (mainExpression.async) {
-      throw new Error('tgpu.fn cannot be async');
-    }
+  let functionNode:
+    | acorn.ArrowFunctionExpression
+    | acorn.FunctionExpression
+    | acorn.FunctionDeclaration
```

Comment:
not sure what's the difference between these expressions vs declarations. couldn't find acorn docs, could you please link it if you have it?

Reply:
FunctionDeclaration nodes are statements, so when we just declare a function, instead of e.g., passing that declaration immediately as an argument. FunctionExpressions are the opposite.

Example of FunctionDeclaration:

```js
function foo() {}
```

Example of FunctionExpression:
```js
bar(function foo() {})
//  ^ here
```